### ISSUE TITLE
Visdiffs: Fix a/b test overrides (logged in and out)

### ISSUE THREAD

**Issue by rachelmcr** [Automattic/wp-e2e-tests#1019]
I'm seeing two problems in a/b test behavior in our visual regression tests:

1) Logged out: Even when the a/b test overrides are set in the signup flow, it isn't always changing the a/b test behavior. From the results I'm seeing this likely only affects the first page in the signup flow, and because it's just testing a visual change it isn't affecting our functional tests.

On the first page in the signup flow, the `siteGoalsShuffle` test still shows a test variant when the control behavior should be set. I see this in our daily test runs (with CircleCI/Sauce Labs) but also in local test runs and manual tests:

![screenshot 2018-03-08 14 10 39](https://user-images.githubusercontent.com/8658164/37157412-b60f537c-22e0-11e8-81d6-af104a823817.png)

2) Logged in: The change made in https://github.com/Automattic/wp-e2e-tests/pull/954 isn't working for some reason when running the visdiff tests on Sauce Labs; the overrides aren't set in the browser. (Local test runs handle the overrides fine.) Error:

```
The localstorage value for AB tests wasn't set correctly.
Expected value is:
'{"domainSuggestionTestV6_20180301":"group_0","upgradePricingDisplayV2_20180305":"original","businessPlanDescriptionAT_20170605":"original","skipThemesSelectionModal_20170904":"show","checklistThankYouForFreeUser_20171204":"hide","checklistThankYouForPaidUser_20171204":"hide","minimizeFreePlan_20180219":"original","minimizedFreePlanForSignedUser_20180308":"original","minimizedFreePlanForNonSignedUser_20180308":"original","siteGoalsShuffle_20180214":"control","redesignedSidebarBanner":"oldBanner"}'
Actual value is:
'{}'
```

You can see this error by setting `{ "sauce": true, "sauceConfig": "osx-chrome" }` in your local config and running the post editor test spec: `./node_modules/.bin/mocha specs/wp-post-editor-spec.js -g '@visdiff'`

**Comment by rachelmcr** [comment]
@alisterscott Do you have any ideas especially for the second (logged in) issue? I'm wondering if it's a timing issue or something else but I'm not sure the best way to debug it.

I wouldn't spend more than 1-2 hours on it (per Lance's suggestion, given that it's low priority at the moment) but figured I'd ask since you worked on that override PR recently.

**Comment by alisterscott** [comment]
Hi @rachelmcr 

For issue 1: I believe this isn't an issue with our tests per se, but rather how that start page reads the AB test values - once it's been read I don't believe it re-renders the page.

So for example, these steps reproduce this manually:

1. As a new non-logged in customer sisit https://wpcalypso.wordpress.com and set the `siteGoalsShuffle` test group to `variant`
2. Visit https://wpcalypso.wordpress.com/start - note that the goals are shuffled ✅ 
3. Change the test group to `control` and refresh/revisit the page - note that goals are still shuffled ❌ 

![set on start page](https://user-images.githubusercontent.com/128826/37189995-e93d0eac-23a2-11e8-97db-6b22daf58cc1.gif)

So, I was thinking, what we can do is instead of setting the AB test groups after we visit the start page (and then refresh) what we can do is visit /log-in, set the AB test groups there, and then visit the start page (and not set them again) - that way the start page will show the correct thing.

I'll put together a new PR to do this now and I'll add you as a reviewer - feel free to merge this if it makes sense and will help you. I'll look at part 2 now.

**Comment by alisterscott** [comment]
For number 2 - these are set on the server and are only able to be overridden when we have the e2e test user agent set (against wordpress.com - wpcalypso.wordpress.com or other environments where we can manually override are ok).

So, some options I can see for number 2 are:

1. use wpcalypso.wordpress.com or
2. override the user agents on saucelabs ([an example](https://stackoverflow.com/questions/28110475/how-to-override-useragent-in-firefox-browser-on-saucelabs)) this might have to be different for different browsers, and may affect the Applitools results?!?

**Comment by alisterscott** [comment]
WRT number 1 - @markryall pointed out that the AB test sets another localstorage item `setSiteGoalsArray` which is why changing the AB test then refreshing the page doesn't change the display.

We _could_ delete that storage item when we set the AB tests on this page, but I'm not sure whether this is a better approach than setting the AB test items _before_ we visit this page:

`localStorage.removeItem( 'setSiteGoalsArray' )`

**Comment by alisterscott** [comment]
I put together another PR to resolve item 1 which simply removes _all_ local storage before setting the AB test overrides - I think this will fix the problem and it's quite simple

@rachelmcr I'll let you choose which solution you'd rather implement 😊

**Comment by rachelmcr** [comment]
Thank you for looking into this! Really appreciate it. :)

For number 1, I missed that there was another localstorage item being set — that definitely explains why refreshing didn't work, and I like the simple approach in #1024 to handle it. That way if there's a future test on the login page with similar behavior, we won't run into the same issue there.

For number 2, thank you for reminding me about the user agent override. Those both sound like reasonable options, so I'll give them a quick try and see what makes the most sense — option 1 is a quicker fix but I worry it will create more maintenance effort over time if changes are made in wpcalypso that don't reach production, so I'm leaning toward option 2.

**Comment by rachelmcr** [comment]
Closing this based on the fixes in #1024 and #1027.
